### PR TITLE
[OASIS-3] Add an OTel collector-contrib module

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

Adds an OTel collector-contrib module for the otel-agent.

This component follows the new component layout described in this RFC: https://docs.google.com/document/d/1sPXUviG-RvMCUvqdwfUYbTOTV05Z_y9MuwlYcyOdG0I/edit

### CODE REVIEW

File: comp/otelcol/collector-contrib/def/component.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package collectorContrib
```
Comment (dineshg13) [inline comp/otelcol/collector-contrib/def/component.go:1]: nit: renme `collectorContrib` - `collectorcontrib`

File: comp/otelcol/collector-contrib/def/component.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+import (
+	"go.opentelemetry.io/collector/otelcol"
```
Comment (dineshg13) [inline comp/otelcol/collector-contrib/def/component.go:4]: can we make it a module ?

Comment (dustmop) [reply comp/otelcol/collector-contrib/def/component.go:4]: Sure thing, pushed a commit to make this available as a go module

File: comp/otelcol/collector-contrib/impl/collectorcontrib.go [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+		azureeventhubreceiver.NewFactory(),
+		azuremonitorreceiver.NewFactory(),
+		bigipreceiver.NewFactory(),
+		carbonreceiver.NewFactory(),
+		chronyreceiver.NewFactory(),
+		cloudflarereceiver.NewFactory(),
+		cloudfoundryreceiver.NewFactory(),
+		collectdreceiver.NewFactory(),
+		couchdbreceiver.NewFactory(),
+		datadogreceiver.NewFactory(),
```
Comment (dineshg13) [inline comp/otelcol/collector-contrib/impl/collectorcontrib.go:236]: Remove datadogreceiver

File: comp/otelcol/collector-contrib/impl/collectorcontrib.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+	processors, err := processor.MakeFactoryMap(processorList...)
+	if err != nil {
+		errs = multierr.Append(errs, err)
+	}
+
+	exporterList := []exporter.Factory{
+		//datadogexporter.NewFactory(),
+		debugexporter.NewFactory(),
+		fileexporter.NewFactory(),
+		otlpexporter.NewFactory(),
```
Comment (dineshg13) [inline comp/otelcol/collector-contrib/impl/collectorcontrib.go:341]: Add OTLP http exporter, signalfx, splunkhec exporter.

**Review by dineshg13 — CHANGES_REQUESTED** [review]

**Review by dineshg13 — APPROVED** [review]

File: comp/otelcol/collector-contrib/impl/go.mod [hunk 5]
```diff
@@ -0,0 +1,580 @@
+module github.com/datadog-agent/DataDog/comp/otelcol/collector-contrib/impl
+
+go 1.21.9
```
Comment (dineshg13) [inline comp/otelcol/collector-contrib/impl/go.mod:3]: Use go 1.21.0

**Review by GustavoCaso — CHANGES_REQUESTED** [review]
I think we should not include the `go.opentelemetry.io/collector/confmap/confmaptest` and `golang.org/x/exp/maps` as part of the rest of the agents.

Any reason why those new imports show as part of the import diff?

**Comment by dustmop** [comment]
> I think we should not include the `go.opentelemetry.io/collector/confmap/confmaptest` and `golang.org/x/exp/maps` as part of the rest of the agents.
> 
> Any reason why those new imports show as part of the import diff?

Looks like the repo depends on `confmaptest` already, seeing this even without this PR:
```
> go mod why go.opentelemetry.io/collector/confmap/confmaptest 
# go.opentelemetry.io/collector/confmap/confmaptest
github.com/DataDog/datadog-agent/comp/otelcol/otlp
go.opentelemetry.io/collector/exporter/loggingexporter
go.opentelemetry.io/collector/exporter/loggingexporter.test
go.opentelemetry.io/collector/confmap/confmaptest
```

Same with `exp/maps`:
```
> go mod why golang.org/x/exp/maps
# golang.org/x/exp/maps
github.com/DataDog/datadog-agent/comp/core/agenttelemetry/agenttelemetryimpl
golang.org/x/exp/maps
```

@GustavoCaso where are you seeing a new addition of those dependencies?

File: comp/otelcol/collector-contrib/def/go.mod [hunk 6]
```diff
@@ -0,0 +1,92 @@
+module github.com/datadog-agent/DataDog/comp/otelcol/collector-contrib/def
+
+go 1.21.9
```
Comment (liustanley) [inline comp/otelcol/collector-contrib/def/go.mod:3]: Can you also use go 1.21.0 here?

Comment (dustmop) [reply comp/otelcol/collector-contrib/def/go.mod:3]: Thanks I missed this one, good catch

**Comment by GustavoCaso** [comment]
> @GustavoCaso where are you seeing a new addition of those dependencies?

**Comment by dustmop** [comment]
Also, important to note that the dependencies being added by this PR are not used by the core agent, or any of the flavors of agent aside from the otel-agent.

**Review by chouetz — APPROVED** [review]
OK for agent-ci-experience modified files.
I let you confirm you did the required changes regarding `replace` directives related to the addition of a new go module.
Thanks

**Review by GustavoCaso — APPROVED** [review]
LGTM

**Review by pducolin — APPROVED** [review]

File: go.mod [hunk 7]
```diff
@@ -214,18 +216,18 @@ require (
 	github.com/netsampler/goflow2 v1.3.3
 	github.com/olekukonko/tablewriter v0.0.5
 	github.com/oliveagle/jsonpath v0.0.0-20180606110733-2e52cf6e6852
-	github.com/open-policy-agent/opa v0.63.0
-	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/resourcetotelemetry v0.97.0 // indirect
+	github.com/open-policy-agent/opa v0.62.0
```
Comment (paulcacheux) [inline go.mod:219]: Why are you downgrading this ?

**Review by paulcacheux — CHANGES_REQUESTED** [review]

File: go.mod (hunk 7 above)
Comment (dustmop) [reply go.mod]: Fixed by the newest commit

File: go.mod [hunk 8]
```diff
@@ -143,11 +145,11 @@ require (
 	github.com/Microsoft/hcsshim v0.12.0-rc.2
 	github.com/acobaugh/osrelease v0.1.0
 	github.com/alecthomas/participle v0.7.1 // indirect
-	github.com/alecthomas/units v0.0.0-20211218093645-b94a6e3cc137
+	github.com/alecthomas/units v0.0.0-20231202071711-9a357b53e9c9
 	github.com/aquasecurity/trivy-db v0.0.0-20231005141211-4fc651f7ac8d
-	github.com/avast/retry-go/v4 v4.6.0
+	github.com/avast/retry-go/v4 v4.5.0
```
Comment (paulcacheux) [inline go.mod:150]: Same here, not sure about the downgrade

**Review by paulcacheux — CHANGES_REQUESTED** [review]

File: go.mod (hunk 8 above)
Comment (dustmop) [reply go.mod]: Another mistake, thanks again for spotting it. I think that's all of them though?

Comment (paulcacheux) [reply go.mod]: I think so yeah

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**